proposals/004-health-checks.md

```diff
@@ -0,0 +1,87 @@
+
+The `RouteController` will be bootstrapped via the `RoutePolicyFactory` where we in the `onInit` callback can register the routes to the `RouteController` which then
+runs as a background thread and orchestrates starting the routes. The `RouteController` will then be responsbible for starting up all these routes, and have support for periodically retry in case a route fails to startup. And only when all routes has been started the `RouteController` has completed starting up all the routes.
+
+We would need to make the discovery and bootstrap of the `RouteController` very easy, which we can do using Camel's existing `RoutePolicyFactory`.
+
+In Camel with Spring Boot we can setup this in the `Main` class via something along the following code lines:
+
+    @Bean
+    RouteController newRouteController(CamelContext camelContext) {
+        return new RouteController(camelContext);
+    }
+
+    @Bean
+    RoutePolicyFactory newRouteController(RouteController controller) {
+        return new RouteControllerRoutePolicyFactory(controller);
+    }
+
+Or:
+
+    @Bean
+    RouteController newRouteController(CamelContext camelContext) {
+        // The route controller is also in charge to add add its own
+        // RoutePolicyFactory to the camel context
+        return new RouteController(camelContext);
+    }
+
+There should be one shared instance of `RouteController` which is given to the `RoutePolicyFactory`, which then creates a `RoutePolicy` for each route that will be controlled.
+
+TODO: We could make `RouteController` an SPI in Apache Camel and provide it OOTB in Apache Camel. Then we can have Camel auto detect if its enabled on startup
+and then automatic set `autoStartup = false`, and then let the `RouteController` startup the routes (We have this idea for Camel 3.0). This can then make it even
+easier to enable. The `RouteController` should have configuration setting to specify how often to retry starting failed routes (we may need to have backoff etc).
+We can then implement a `DefaultRouteController` in camel-core, and then make it easy to use by setting `CamelContext.setRouteControllerEnabled(true)` (find a good name). With Camel on Spring Boot then its a matter of setting this in the `application.properties` via `camel.springboot.route-controller.enabled = true`.
+
+Besides starting up routes the `RouteController` should have a Java, JMX API, and REST which can report back status of the running integrations (i.e. running routes).
+At minimum it should collect status of each route with details:
+
+- route id
+- current status (starting, started, stopping, stopped)
+- health (unknown, ok, failure)
+- failure error message (exception message + stacktrace)
+
+TODO: API details to be determined
+
```

Comment:
Question: The RouteController can start routes. Is this the only lifecycle operation which is should support ? Or should the be a way to suspend a route and restart it ? Maybe even auto-suspend when a certain metrics threshold has been exceeded. 

I'm asking because this interacts with the Integration lifecycle. It might be nice for the user to _pause_ an integration (maybe then by contacting the RouteController ?) and restart it later without a complete undeploy / redeploy (so the integration continues to run), so that this action is much quicker. // @kcbabo this is one of the points I'd like to discuss for the intergration lifecycle.